File: worker/app/blueprint_contracts/plan_path.py
"""PlanPath provenance contract twin (Phase 1)."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal, Optional, TypedDict

PlanPathType = Literal[
    "report_page", "blueprint_sheet", "spec_section", "user_entry", "derived"
]


@dataclass
class BlueprintPlanPath:
    import_session_id: str
    path_type: PlanPathType
    confidence: float = 0.0
    source_document_id: Optional[str] = None
    file_name: Optional[str] = None
    document_type: Optional[str] = None
    provider: Optional[str] = None
    page_number: Optional[int] = None
    section_label: Optional[str] = None
    table_label: Optional[str] = None
    diagram_label: Optional[str] = None
    source_text_excerpt: Optional[str] = None
    source_coordinates: Optional[dict[str, Any]] = None
    id: Optional[str] = None


class PlanPathValidation(TypedDict, total=False):
    ok: bool
    reason: str


def validate_plan_path_present(plan_path: Optional[dict[str, Any] | BlueprintPlanPath]) -> PlanPathValidation:
    if plan_path is None:
        return {"ok": False, "reason": "plan_path is missing"}
    if isinstance(plan_path, BlueprintPlanPath):
        pp = plan_path.__dict__
    else:
        pp = plan_path
    if not pp.get("path_type"):
        return {"ok": False, "reason": "plan_path.path_type is required"}
    has_anchor = any(
        pp.get(k)
        for k in (
            "source_document_id",
            "file_name",
            "section_label",
            "table_label",
            "diagram_label",
            "source_text_excerpt",
        )
    ) or isinstance(pp.get("page_number"), int)
    if not has_anchor:
        return {
            "ok": False,
            "reason": "plan_path needs at least one anchor (source_document_id, page_number, section_label, table_label, diagram_label, source_text_excerpt, or file_name)",
        }
    confidence = pp.get("confidence")
    if isinstance(confidence, (int, float)) and (confidence < 0 or confidence > 1):
        return {"ok": False, "reason": "plan_path.confidence must be in [0,1]"}
    return {"ok": True}


def requires_plan_path(_trade_id: str) -> bool:
    return True

File: supabase/functions/_shared/blueprint-importer/plan-path.ts
// Blueprint Importer v2 — PlanPath provenance contract (Phase 1).
// Side-effect-free.

export type PlanPathType =
  | "report_page"
  | "blueprint_sheet"
  | "spec_section"
  | "user_entry"
  | "derived";

export interface BlueprintPlanPath {
  id?: string;
  import_session_id: string;
  source_document_id?: string | null;
  path_type: PlanPathType;
  file_name?: string | null;
  document_type?: string | null;
  provider?: string | null;
  page_number?: number | null;
  section_label?: string | null;
  table_label?: string | null;
  diagram_label?: string | null;
  source_text_excerpt?: string | null;
  source_coordinates?: Record<string, unknown> | null;
  confidence: number; // 0..1
}

/**
 * Phase 0 rule: every auto-populated material/labor line must carry a non-empty
 * PlanPath. This is the validator. Returns null when valid, or a reason string.
 */
export function validatePlanPathPresent(
  plan_path: Partial<BlueprintPlanPath> | null | undefined,
): { ok: true } | { ok: false; reason: string } {
  if (!plan_path) return { ok: false, reason: "plan_path is missing" };
  if (!plan_path.path_type) return { ok: false, reason: "plan_path.path_type is required" };
  const hasAnchor =
    !!plan_path.source_document_id ||
    !!plan_path.file_name ||
    !!plan_path.section_label ||
    !!plan_path.table_label ||
    !!plan_path.diagram_label ||
    !!plan_path.source_text_excerpt ||
    typeof plan_path.page_number === "number";
  if (!hasAnchor) {
    return { ok: false, reason: "plan_path needs at least one anchor (source_document_id, page_number, section_label, table_label, diagram_label, source_text_excerpt, or file_name)" };
  }
  if (typeof plan_path.confidence === "number" && (plan_path.confidence < 0 || plan_path.confidence > 1)) {
    return { ok: false, reason: "plan_path.confidence must be in [0,1]" };
  }
  return { ok: true };
}

/**
 * Trades that MUST carry a PlanPath on every measurement/draft line they emit.
 * Currently: all MVP-supported trades. Measurement-object-only trades surface
 * via measurement objects which themselves require PlanPaths.
 */
export function requiresPlanPath(_trade_id: string): boolean {
  return true;
}
